docs/source/design/input_output_specifications.md

```diff
@@ -0,0 +1,692 @@
+    validation_schema : Optional[Dict[str, Any]] = None # Optional schema for validator
+    description : Optional[str] = None                # Brief description
+
+    @property
+    def key(self) -> str:
+        """Canonical identifier used everywhere else."""
+        return f"{self.formats[0]}{self.modality}:{self.semantic}:{self.version}"
+```
+
+
+### `DatasetSpec`
+
+
+```python
+class DatasetSpec(BaseModel, extra="forbid"):
+    """
+    Describes a dataset's structure, provenance, and associated specifications.
+    """
+    name         : str                                 # Human-readable dataset name
+    description  : Optional[str] = None                # Brief description of the dataset
+    specs        : List[InputOutputSpecs]           # Unique set of input/output specifications
+    provenance   : Optional[Provenance] = None         # Dataset origin and authorship info
+    checksum     : Optional[str] = None                # Optional data integrity checksum
+    created_at   : Optional[datetime] = None           # Creation timestamp
+    metadata     : Dict[str, Any] = {}                 # Arbitrary additional metadata
+```
+
+
+###  `ModelSpec`
+
+
+```python
+class ModelSpec(BaseModel, extra="forbid"):
+    """
+    Describes a runnable model, its inputs, outputs, and metadata.
+    """
+    name              : str                                 # Human-readable model name
+    description       : Optional[str] = None                # Brief description of the model
+    version           : str                                 # Model version
+    required_inputs   : List[InputOutputSpecs] = []      # All required input specifications
+    optional_inputs   : List[InputOutputSpecs] = []      # Optional input specifications (e.g., masks)
+    outputs           :  List[InputOutputSpecs]  
+    compute           : Resource = Resource()               # Compute resource requirements
+    provenance        : Optional[Provenance] = None         # Model origin and authorship info
+    metadata          : Dict[str, Any] = {}                 # Arbitrary additional metadata
+    supported_tasks   : Optional[List[str]] = None          # List of supported TaskSpec names/IDs
+    supported_metrics : Optional[List[str]] = None          # List of supported MetricSpec names/IDs
+```
+
+
+
+### `MetricSpec`
+
+```python
+class MetricSpec(BaseModel, extra="forbid"):
+    """
+    Describes a metric used to evaluate model performance.
+    """
+    name        : str                                 # Human-readable metric name
+    description : Optional[str] = None                # Brief description of the metric
+    category    : str                                 # Metric type/category (e.g., "clustering", "classification")
+    impl        : str                                 # Python function implementing the metric (python:module:function)
+    direction   : Optional[Literal["higher_better", "lower_better"]] = None
```

Comment:
Usually this is formulated as an objective with the results being "maximize" or "minimize".